### PR TITLE
Examples for some of the near-sdk functions

### LINKED ISSUE #764: [EXT] Ext Contract Description and Example
https://docs.rs/near-sdk/4.0.0-pre.7/near_sdk/attr.ext_contract.html

This function is missing both a descriptive sentence and an example.

Examples should be simple, easy-to-follow, and useful to make onboarding and learning Rust SDK as straightforward as possible. See array primitive for examples of code snippet examples: https://doc.rust-lang.org/nightly/core/primitive.array.html), https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#784-790 for examples in code.

### PR DESCRIPTION
**Pull request by NinoLipartiia4ire**
Closes #764, #765, #766, #767, #768, #769

Examples for macros (`near_bindgen`, `init` and `exc_contract`) are marked with `ignore`, that is because in order to test them we would need to import near_sdk to near_sdk_macros, which is impossible because of dependency cycle.

### CODE REVIEW

File: near-sdk-macros/src/lib.rs [hunk 1]
```diff
@@ -11,6 +11,25 @@ use quote::quote;
 use syn::visit::Visit;
 use syn::{File, ItemEnum, ItemImpl, ItemStruct, ItemTrait};
 
+/// This macro is used on a struct and the function implementations
+/// to generate the necessary code to be a valid NEAR contract and expose
+/// the intended functions to be able to be called externally.
+///
+/// # Examples
+///
+/// ```ignore
```
Comment (austinabell) [inline near-sdk-macros/src/lib.rs:20]:
```suggestion
/// ```no_run
```

Why not compile it to make sure docs stay up to date?

File: near-sdk-macros/src/lib.rs [hunk 2]
```diff
@@ -45,6 +64,19 @@ pub fn near_bindgen(_attr: TokenStream, item: TokenStream) -> TokenStream {
     }
 }
 
+/// `ext_contract` takes a Rust Trait and converts it to a module with static methods.
+/// Each of these static methods takes positional arguments defined by the Trait,
+/// then the receiver_id, the attached deposit and the amount of gas and returns a new Promise.
+///
+/// # Examples
+///
+/// ```ignore
```
Comment (austinabell) [inline near-sdk-macros/src/lib.rs:73]:
```suggestion
/// ```no_run
```
ditto

Comment (austinabell) [inline near-sdk-macros/src/lib.rs:73]:
```suggestion
/// ```no_run
```
ditto

File: near-sdk-macros/src/lib.rs [hunk 3]
```diff
@@ -108,6 +140,27 @@ pub fn result_serializer(_attr: TokenStream, item: TokenStream) -> TokenStream {
...
+/// #[near_bindgen]
+/// impl Contract {
+///     #[init]
+///     pub fn new(initial_data: i8) -> Self {
+///         Self {
+///             data: initial_data,
+///         }
+///     }
+/// }
+/// ```
```
Comment (austinabell) [inline near-sdk-macros/src/lib.rs:163]:
```suggestion
```

No need for this, it's a deprecated annotation and will be removed

Comment (austinabell) [inline near-sdk-macros/src/lib.rs:163]:
```suggestion
```

No need for this, it's a deprecated annotation and will be removed

File: near-sdk/src/environment/env.rs [hunk 4]
```diff
@@ -259,6 +259,18 @@ pub fn random_seed_array() -> [u8; 32] {
...
 /// Hashes the random sequence of bytes using sha256.
+///
+/// # Examples
+///
+/// ```
+/// use near_sdk::env::sha256;
+/// use hex;
+///
+/// assert_eq!(
+///     sha256("The phrase that will be hashed".as_bytes()),
```
Comment (austinabell) [inline near-sdk/src/environment/env.rs:270]:
```suggestion
///     sha256(b"The phrase that will be hashed"),
```
Simplicity

Comment (austinabell) [inline near-sdk/src/environment/env.rs:270]:
```suggestion
///     sha256(b"The phrase that will be hashed"),
```
Simplicity

File: near-sdk/src/environment/env.rs [hunk 5]
```diff
@@ -639,6 +651,16 @@ pub fn log(message: &[u8]) {
...
 /// If another key-value existed in the storage with the same key it returns `true`, otherwise `false`.
+///
+/// # Examples
+///
+/// ```
+/// use near_sdk::env::{storage_write, storage_read};
+///
+/// assert!(!storage_write("key".as_bytes(), "value".as_bytes()));
+/// assert!(storage_write("key".as_bytes(), "another_value".as_bytes()));
+/// assert_eq!(storage_read("key".as_bytes()).unwrap(), "another_value".as_bytes());
```
Comment (austinabell) [inline near-sdk/src/environment/env.rs:662]:
```suggestion
/// assert!(!storage_write(b"key", b"value"));
/// assert!(storage_write(b"key", b"another_value"));
/// assert_eq!(storage_read(b"key").unwrap(), b"another_value");
```

Comment (austinabell) [inline near-sdk/src/environment/env.rs:662]:
```suggestion
/// assert!(!storage_write(b"key", b"value"));
/// assert!(storage_write(b"key", b"another_value"));
/// assert_eq!(storage_read(b"key").unwrap(), b"another_value");
```

File: near-sdk/src/environment/env.rs [hunk 6]
```diff
@@ -655,6 +677,16 @@ pub fn storage_write(key: &[u8], value: &[u8]) -> bool {
...
 /// Reads the value stored under the given key.
+///
+/// # Examples
+///
+/// ```
+/// use near_sdk::env::{storage_write, storage_read};
+///
+/// assert!(storage_read("key".as_bytes()).is_none());
+/// storage_write("key".as_bytes(), "value".as_bytes());
+/// assert_eq!(storage_read("key".as_bytes()).unwrap(), "value".as_bytes());
```
Comment (austinabell) [inline near-sdk/src/environment/env.rs:688]: can you swap the as_bytes for byte strings here too? `b".."`

Comment (austinabell) [inline near-sdk/src/environment/env.rs:688]: can you swap the as_bytes for byte strings here too? `b".."`

**Review by austinabell — CHANGES_REQUESTED** [review]

**Comment by austinabell** [comment]
@NinoLipartiia4ire seems like #782 does a similar function but this came before. If you make the adjustments in the PR I can pull this in, or if you don't have the bandwidth let me know and I can get this across the line for you :)

File: near-sdk-macros/src/lib.rs (hunk 1 above)
Comment (NinoLipartiia4ire) [reply near-sdk-macros/src/lib.rs:20]: I think this would have the same problem, the test would fail because of dependency cycle.

**Comment by NinoLipartiia4ire** [comment]
@austinabell I've made these adjustments, except for replacing ignore with no_run

**Comment by austinabell** [comment]
> @austinabell I've made these adjustments, except for replacing ignore with no_run, because this would have had the same problem

I've modified the near_bindgen docs and rebuilt the wasm files. https://github.com/near/near-sdk-rs/blob/master/CONTRIBUTING.md has the information for doing all this, if you plan to do another PR in the future :)

**Review by austinabell — APPROVED** [review]

**Pull request merged**